presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,513 @@
+
+    @Description("Returns length of the geometry if it is a linestring or multilinestring")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stLength(@SqlType(GEOMETRY) Slice input)
+    {
+        return deserialize(input).getEsriGeometry().calculateLength2D();
+    }
+
+    @Description("Returns the maximum X coordinate of geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stMaxX(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        Envelope envelope = getEnvelope(geometry);
+        return envelope.getXMax();
+    }
+
+    @Description("Returns the maximum Y coordinate of geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stMaxY(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        Envelope envelope = getEnvelope(geometry);
+        return envelope.getYMax();
+    }
+
+    @Description("Returns the minimum X coordinate of geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stMinX(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        Envelope envelope = getEnvelope(geometry);
+        return envelope.getXMin();
+    }
+
+    @Description("Returns the minimum Y coordinate of geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stMinY(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        Envelope envelope = getEnvelope(geometry);
+        return envelope.getYMin();
+    }
+
+    @Description("Returns the number of interior rings in the polygon")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long stInteriorRingNumber(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "st_interior_ring_number", POLYGON);
+        return ((OGCPolygon) geometry).numInteriorRing();
+    }
+
+    @Description("Returns the number of points in the geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long stPointNumber(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        if (geometry.geometryType().equals(POLYGON)) {
+            Polygon polygon = (Polygon) geometry.getEsriGeometry();
+            return polygon.getPointCount() + polygon.getPathCount();
+        }
+        else if (geometry.geometryType().equals(POINT)) {
+            return geometry.getEsriGeometry().isEmpty() ? 0 : 1;
+        }
+        else if (geometry.geometryType().equals(MULTIPOINT)) {
+            return ((MultiPoint) geometry.getEsriGeometry()).getPointCount();
+        }
+        return ((MultiPath) geometry.getEsriGeometry()).getPointCount();
+    }
+
+    @SqlNullable
+    @Description("Returns true if and only if the line is closed and simple")
```

Comment:
What does closed and simple mean?

Reply:
simple means:
Each figure of the instance must not intersect itself, except at its endpoints.
No two figures of the instance can intersect each other at a point that is not in both of their boundaries.